src/main/java/org/opensearch/neuralsearch/query/HybridScoreBlockBoundaryPropagator.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Scorer;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Objects;
+
+/**
+ * This class functions as a utility for propagating block boundaries within disjunctions.
```

Comment:
What is a block boundary?

Reply:
There is a logic in Lucene that iterates on range of docs, e.g. [DocIdIterator.range](https://github.com/apache/lucene/blob/branch_9_10/lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java#L96). For that doc ids are split into intervals and we need to know where in those intervals our range starts and/or ends. I understand that block is a pieces of range that is part of one interval, and boundary is the doc id that marks start/end of that block. 
This class is actually inspired by [similar class in Lucene](https://github.com/apache/lucene/blob/branch_9_10/lucene/core/src/java/org/apache/lucene/search/DisjunctionScoreBlockBoundaryPropagator.java), we just cannot call it directly as it's not public

Reply:
Essentially, term postings (matching doc IDs) are grouped into blocks of docs (I forget the number) and the start of the block holds some metadata ("impacts") including maximum term frequency of the term in any doc in the block and the minimum "norm" (doc length of a matching doc for any of the terms) for any doc in the block.

Using those values gives you a (potentially loose, if the max term frequency and minimum norm don't come from the same doc) upper bound on the possible score contribution of a doc in the block. Once you compute that across all terms, you can compute another (possibly even looser) upper bound on all scores for docs in this block. If that upper bound is below the score of the Nth hit when you're collecting the top N, you know you can safely skip the block -- none of its docs could possibly be good enough to win the title of TopDocs.